### ISSUE TITLE
[UX] Remove the scary confirmation form for 'Reset display mode'

### ISSUE THREAD

**Issue by BWPanda** [backdrop/backdrop-issues#3036]
1. Create a new content type (preferably on a site with other existing content types)
2. Go to the 'Manage Display's tab/page
3. Note the presence of a Teaser (which we have on other content types, but don't want on this one)
4. Use the drop-down menu next to Teaser to select 'Reset to default' (to remove it)
5. Read the text on the conformation page:

> The Teaser display mode has been customized in the following places:
>
> - Page
> - Post
> - Other content type
> - This content type
> - Etc.
>
> A reset will cause all customizations to be lost. The Default display mode will be used instead.

6. Freak out because you don't want to remove the Teaser (or your customisations to it) from your other content types.
7. Hesitantly click 'Reset' and cross your fingers

This, in my opinion, is very confusing and unnecessary. Why list other content types at all if they're not affected? Or are they affected!?

**Comment by BWPanda** [comment]
Ok, so I've just been reading through https://github.com/backdrop/backdrop-issues/issues/315 which is where this all came from. I can now see the purpose of the text on the confirmation page, however there are a few issues I think we need to fix:

### Rename 'Reset to default' to 'Disable'

The 'Reset to default' button seems to simply disable the view mode for that content type (both core view modes (Teaser, etc.) and custom ones). If there's a button to 'Enable' view modes, the corresponding button should be labelled 'Disable'.

### Change the 'reset' wording

The reset confirmation currently says:

> A reset will cause all customizations to be lost. The Default display mode will be used instead.

This is misleading, as I can disable a view mode, then enable it again, and all the fields/formats are still the same (i.e. no customisations were lost). I think we can simplify the wording to:

> A reset will cause the Default display mode to be used instead.

### Remove the list of content types from reset confirmation pages

When you reset a display mode, is just gets disabled. Even if it's used on other content types, they're not affected. So there's no point showing the list of content types where it's used on the *reset* confirmation page. This is just confusing and causes some unnecessary worry (as per the OP above).

**Comment by BWPanda** [comment]
Here're some screenshots for how it now looks:

![image](https://user-images.githubusercontent.com/2385329/105131911-02529a80-5b3e-11eb-93f5-7aa9c3a8dcaa.png)

![image](https://user-images.githubusercontent.com/2385329/105131936-0da5c600-5b3e-11eb-840f-021a14e4c9d1.png)

![image](https://user-images.githubusercontent.com/2385329/105131953-15656a80-5b3e-11eb-8b4a-a1c7bd82a53f.png)
 ^ This last screenshot is unchanged, but shows that the list of content types is still displayed on Delete confirmation pages.

**Comment by stpaultim** [comment]
Based on your screenshots, this change seems to make sense and the new wording seems sensible.

**Comment by klonos** [comment]
I believe that we should wait for #3402 to be merged before we start looking at this, but here's some feedback anyway...

>If there's a button to 'Enable' view modes, the corresponding button should be labelled 'Disable'.

Reading this enabled/disabled argument you make @BWPanda, the suggested wording changes make sense to me. **However**, now looking at the display mode overview page, the "Customized for %type" and "Not customized for %type" headings in the table start not making complete sense. I start thinking that instead of `Enable`/`Disable` perhaps it'd make sense to have the labels of these buttons be `Customize`/`Reset` (in others words, I'm suggesting that the "reset" word being used originally wasn't the problem - "enable" was) 🤔 

> This is misleading, as I can disable a view mode, then enable it again, and all the fields/formats are still the same (i.e. no customisations were lost).

Interesting 🤔 ...while a display mode is "disabled" for a content type, does it mean that it cannot be used? ...or that it can, but with its settings configured same as the _Default_ display? I believe the latter. For instance, see the available options when selecting a display mode for a view that shows content:

![image](https://user-images.githubusercontent.com/2423362/105140111-ed571680-5aff-11eb-942b-101d172f1d89.png)

If it is the latter, then it is indeed not "disabled", rather than "reset" (or rather falling back to the default display mode _settings_ if you want to be _technically_ accurate). Right?

PS: in the "Manage displays" tab, we refer to these things as both "displays" as well as "display modes", which is also a bit confusing. So the label of the tab, and the operation button refers to them as "displays"; whereas the "+ Add ..." link, the help text, and the table heading all use "display modes". We should fix that.

PPS: given that the _Default_ display mode is the most commonly-used one, still this table follows a different pattern than the layouts table, where the defaults/fall-backs are listed at the bottom. I realize that this is a different scenario, but from a consistency point of view, just saying 🤷

**Comment by klonos** [comment]
...here's more thoughts (sorry):

> I can disable a view mode, then enable it again, and all the fields/formats are still the same (i.e. no customisations were lost).

If that is true, then this action is not as destructive (think disabling a view for example), so perhaps we shouldn't be showing a confirmation message at all here. We can "disable" (or "revert"), and then just show the "The display mode Teaser will now use the Default settings." as we currently do.

What I see as an improved UX workflow would be to show something like this when _deleting_ a display mode:

![image](https://user-images.githubusercontent.com/2423362/105143286-4e80e900-5b04-11eb-982b-2cf59195c818.png)
(quick mockup - I haven't thought wording much)

**Comment by BWPanda** [comment]
> I believe that we should wait for #3402 to be merged before we start looking at this

@klonos Why is that? This PR doesn't change anything that that PR affects...

Regarding all the changes mentioned, I can see this issue getting really complicated. But thanks for your feedback @klonos, especially the "see the available options when selecting a display mode for a view that shows content" part - I hadn't realised that!

Since the purpose of this issue was to prevent the scary moment when disabling the Teaser view mode (thinking it was going to disable *all* the Teaser view modes), what I'm going to do is focus on that part in this PR, then open others for everything else we've discovered here...

**Comment by BWPanda** [comment]
I've updated the PR to simply remove the confirmation form when resetting display modes. Resetting now works the same (but opposite) as enabling.

**Comment by BWPanda** [comment]
Woohoo, tests passing! (can someone please check I didn't just disable the relevent ones somehow...?)

**Comment by klonos** [comment]
Thanks for breaking my feedback into discrete issues @BWPanda 🙏 

>> I believe that we should wait for #3402 to be merged before we start looking at this
>
> @klonos Why is that? This PR doesn't change anything that that PR affects...

Not really relevant now that you've created #4890, but in the PR for #3402 the machine name gets moved from the dedicated column into the "Display modes" column, making room for a "Description" column:

![image](https://user-images.githubusercontent.com/2423362/105193927-bace0d80-5b41-11eb-8d4e-3aeaf2ec00a0.png)

...the description for the non-customized display modes is "Uses the Default display mode" (nice btw 👍), which would make the change from "Enable" to "Customize" more relevant/obvious, or would at least make more sense.

> Since the purpose of this issue was to prevent the scary moment ...I've updated the PR to simply remove the confirmation form when resetting display modes. Resetting now works the same (but opposite) as enabling.

Yup, I think that is much better now 👍

**Comment by jenlampton** [comment]
I find the current confirmation form on these really helpful. I always want to disable a display mode, forgetting that It will disable it on all the other node types too. Having a scary confirm form always stops me from doing something stupid (and breaking my other content types).

> Why list other content types at all if they're not affected? Or are they affected!?

They **are** affected... which is why the message is so scary. It needs to stop me from doing something I didn't intend. 

Every time I see the scary message, I think about creating an issue to fix that, but somehow forget.

> Even if it's used on other content types, they're not affected.

Are you sure? Maybe this was fixed already, but it would be good to find the issue where that change happened to confirm.

**Comment by BWPanda** [comment]
@jenlampton This must have been fixed, because I'm not seeing any indication that resetting a display mode for one content type affects any others...

![Peek 2021-01-21 09-34](https://user-images.githubusercontent.com/2385329/105249204-d8e05000-5bcb-11eb-9dce-f09bb39751ab.gif)

**Comment by jenlampton** [comment]
I haven't found any issues for it specifically. I'm wondering if it was fixed by our handling of config for display modes? Or maybe entity_view_mode in core? still looking...

**Comment by BWPanda** [comment]
Maybe it's a bug (a good one!) 😄

**Comment by BWPanda** [comment]
From what I can see @jenlampton, this has always been the case ever since the Entity View Mode module was added to core: https://github.com/backdrop/backdrop/commit/4e4496455d9133d63fd20830fa739b403eac0a5b

For example:
```
public function getActualViewMode($entity_type, $bundle, $view_mode) {
  $view_mode_settings = field_view_mode_settings($entity_type, $bundle);
  return (!empty($view_mode_settings[$view_mode]['custom_settings']) ? $view_mode : 'default');
}
```
(source: https://github.com/backdrop/backdrop/commit/4e4496455d9133d63fd20830fa739b403eac0a5b#diff-70848d46bcf39d81e299cb1c09337d8bd50ce7fe97d699918eb49738fd4b932aR142-R145)

The `custom_settings` setting (customised or not customised) is set in `$view_mode_settings`, which is obtained from the `$entity_type` and `$bundle`. So this setting has always been tried to the bundle, and therefore resetting a display mode for one bundle has never affected any others (from what I've seen in the code anyway).

**Comment by jenlampton** [comment]
Yeah, IIRC there was a problem with config for view modes / field display settings back in the early days (way before entity_view_mode) that resulted in a reshuffling of how these things were stored. I'm thinking perhaps that's what fixed it (or maybe that was what broke it?). But I can't find that issue. Anyway, I tested in Drupal 7 core and in Drupal 7 entity_view_mode, and both also saved settings for each entity/bundle. So whatever caused us to list each bundle here is lost to me.  Anywho -- RTBC again :) and nice improvement!!!

**Comment by BWPanda** [comment]
Can I go ahead and commit this then (myself, @klonos and @jenlampton all agree it's RTBC), or do we need more people/time before it can be committed...?

**Comment by jenlampton** [comment]
It's up to you whether you're confident in committing a PR that you wrote, or if you want another committer (or non committer!) to review it. @quicksketch usually errs on the site of caution. I usually err on the side of progress (but I break a lot more things than he does) heh.

**Comment by BWPanda** [comment]
Fair enough, I'll let another core committer do the honours.

In the meantime, what's the thought on whether this should be in the next bug-fix release or not?

**Comment by jenlampton** [comment]
I would think it's a good bug-fix release. One tiny string change `Reset to default` becomes `Reset Display mode`.

If that were `Reset` instead it wouldn't require a translation update... ;)

**Comment by BWPanda** [comment]
Unfortunately I think this one string change is the least of the translators problems... I think https://github.com/backdrop/backdrop-issues/issues/4890 and.or https://github.com/backdrop/backdrop-issues/issues/4891 will big a bigger headache for them...

Because of those other issues (particularly https://github.com/backdrop/backdrop-issues/issues/4891), I don't think changing the text to 'Reset' here would be worthwhile, since it'll likely be changed in that issue instead/as well.

**Comment by klonos** [comment]
...plus `Reset Display mode` should not capitalize the "D" in "display".

**Comment by jenlampton** [comment]
If we use `Reset` here than https://github.com/backdrop/backdrop-issues/issues/4891 won't affect this, since that only changes `Displays` to `Display modes` :)

**Comment by jenlampton** [comment]
Changing to NW due to the text change on `Reset Display mode`

**Comment by BWPanda** [comment]
PR updated.

**Comment by BWPanda** [comment]
Thanks @klonos, @stpaultim and @jenlampton for help with this. I've merged https://github.com/backdrop/backdrop/pull/3505 into 1.x and 1.18.x.

**Comment by quicksketch** [comment]
Sorry I didn't get to this in time. In reviewing the PR, I found that it introduces a security risk. We can't switch from a POST request to a GET request without also protecting the URL with a token. Using a form gives us natural protection against CSRF attacks because they have form tokens built-in. If using a GET request, any site could cause a logged-in administrator to make a request to their Backdrop site and disable text formats.

For now I've reverted the commits from the PR: https://github.com/backdrop/backdrop/pull/3505

**Comment by quicksketch** [comment]
See `system_theme_enable()` or `comment_approve()` for examples of how tokens are checked on GET requests.

It usually involved creating a token an appending it to the URL, then checking it in the menu call back like this:
```
  if (!isset($_GET['token']) || !backdrop_valid_token($_GET['token'], $display_name)) {
    return MENU_ACCESS_DENIED;
  }
```

**Comment by markabur** [comment]
I went through the exact scenario in the OP and eventually found this issue. I think it would be good to remove the false and misleading information from the confirmation screen. I don't have an opinion about removing the confirmation screen altogether, or about what the buttons are named. On my site I am simply hacking core like this and it is a big improvement:

```
  // Check to see if this display mode is in use on any other bundles.
//  $active = array();
//  $entity_info = entity_get_info($entity_type);
//  foreach ($entity_info['bundles'] as $machine => $bundle) {
//    $settings = field_view_mode_settings($entity_type, $machine);
//    if (isset($settings[$view_mode_name]) && $settings[$view_mode_name]) {
//      $active[] = t($bundle['label']);
//    }
//  }
//
//  if (!empty($active)) {
//    $message .= '<p>' . t('The %view-mode display mode has been customized in the following places:', array('%view-mode' => $view_mode['label'])) . '</p>';
//    $message .= theme('item_list', array('items' => $active));
//  }

```
